File: android/app/src/main/java/com/example/mobileuiautomationdemo/WeatherClient.kt
package com.example.mobileuiautomationdemo

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

object WeatherClient {
    private const val BASE_URL = "https://api.open-meteo.com/v1/forecast"
    private const val CITY = "San Francisco"
    private const val LATITUDE = 37.7749
    private const val LONGITUDE = -122.4194

    suspend fun fetchCurrentTemperatureFahrenheit(): String = withContext(Dispatchers.IO) {
        val url = URL(
            "$BASE_URL?latitude=$LATITUDE&longitude=$LONGITUDE" +
                "&current=temperature_2m&temperature_unit=fahrenheit"
        )
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 15_000
        }

        try {
            val responseCode = connection.responseCode
            if (responseCode != HttpURLConnection.HTTP_OK) {
                throw WeatherException("HTTP $responseCode")
            }

            val body = connection.inputStream.bufferedReader().use { it.readText() }
            val current = JSONObject(body).getJSONObject("current")
            val temperature = current.getDouble("temperature_2m")
            "$CITY: ${formatTemperature(temperature)}°F"
        } finally {
            connection.disconnect()
        }
    }

    private fun formatTemperature(value: Double): String {
        return if (value % 1.0 == 0.0) {
            value.toInt().toString()
        } else {
            String.format("%.1f", value)
        }
    }
}

class WeatherException(message: String) : Exception(message)

File: ios/DemoApp/WeatherClient.swift
import Foundation

enum WeatherClientError: Error {
    case invalidResponse
    case missingTemperature
}

enum WeatherClient {
    private static let baseURL = "https://api.open-meteo.com/v1/forecast"
    private static let city = "San Francisco"
    private static let latitude = 37.7749
    private static let longitude = -122.4194

    static func fetchCurrentTemperatureFahrenheit() async throws -> String {
        var components = URLComponents(string: baseURL)
        components?.queryItems = [
            URLQueryItem(name: "latitude", value: String(latitude)),
            URLQueryItem(name: "longitude", value: String(longitude)),
            URLQueryItem(name: "current", value: "temperature_2m"),
            URLQueryItem(name: "temperature_unit", value: "fahrenheit"),
        ]

        guard let url = components?.url else {
            throw WeatherClientError.invalidResponse
        }

        let (data, response) = try await URLSession.shared.data(from: url)
        guard let httpResponse = response as? HTTPURLResponse,
              (200 ... 299).contains(httpResponse.statusCode) else {
            throw WeatherClientError.invalidResponse
        }

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let current = json?["current"] as? [String: Any]
        guard let temperature = current?["temperature_2m"] as? Double else {
            throw WeatherClientError.missingTemperature
        }

        return "\(city): \(formatTemperature(temperature))°F"
    }

    private static func formatTemperature(_ value: Double) -> String {
        if value.truncatingRemainder(dividingBy: 1) == 0 {
            return String(Int(value))
        }
        return String(format: "%.1f", value)
    }
}
